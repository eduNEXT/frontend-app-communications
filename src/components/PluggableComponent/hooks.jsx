import React, { useState, useEffect, useRef } from 'react';
import loadable from '@loadable/component';
import useDeepCompareEffect from 'use-deep-compare-effect';

import { isPluginAvailable, getPluginsByPrefix } from './utils';

/**
 * Component that dynamically loads and displays a list of plugins.
 *
 * @param {Object} props - The component props.
 * @param {React.ReactNode} props.children - Children components to be passed to each plugin.
 * @param {Array} props.plugins - An array of plugin objects to be loaded.
 * @param {React.ReactNode} props.loadingComponent - Component to display while each plugin is loading.
 * @param {Object} props.pluggableComponentProps - Additional props to be passed to each dynamically loaded plugin.
 * @returns {React.ReactNode} The rendered component.
 */
export const usePlugins = (plugins, pluggableComponentProps, prefix, loadingComponent) => {
  const [pluginComponents, setPluginComponents] = useState({});
  const loadedAllPluginsRef = useRef(null);

  useEffect(() => {
    const loadPlugins = (pluginsList) => {
      pluginsList.forEach((plugin, index) => {
        // Initially set the loading component for each plugin
        setPluginComponents(previousPluginComponents => ({
          ...previousPluginComponents,
          [plugin.id]: loadingComponent || null,
        }));

        const loadPlugin = async () => {
          try {
            const hasModuleInstalled = await isPluginAvailable(plugin.name);
            if (hasModuleInstalled) {
              const PluginComponent = loadable(() => import(`@node_modules/@openedx-plugins/${plugin.name}`));
              setPluginComponents(previousPluginComponents => ({
                ...previousPluginComponents,
                [plugin.id]: (
                  <PluginComponent {...pluggableComponentProps} />
                ),
              }));
            }
          } catch (error) {
            console.error(`Failed to load plugin ${plugin.name}:`, error);
            // Set to null in case of an error
            setPluginComponents(previousPluginComponents => ({
              ...previousPluginComponents,
              [plugin.id]: null,
            }));
          } finally {
            const isLastPlugin = index === pluginsList.length - 1;
            if (isLastPlugin) {
              loadedAllPluginsRef.current = true;
            }
          }
        };

        loadPlugin();
      });
    };

    const pluginsToLoad = prefix ? getPluginsByPrefix(prefix) : plugins;

    if (pluginsToLoad.length) {
      loadPlugins(pluginsToLoad);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useDeepCompareEffect(() => {
    const updatePluginsWithNewProps = () => {
      const updatedComponents = Object.keys(pluginComponents).reduce((previousPluginComponents, pluginKey) => {
        const PluginComponent = pluginComponents[pluginKey];
        // Check if the component is a valid React element and not a loading or error state
        if (React.isValidElement(PluginComponent)) {
          const UpdatedComponent = React.cloneElement(PluginComponent, pluggableComponentProps);
          return {
            ...previousPluginComponents,
            [pluginKey]: UpdatedComponent,
          };
        }
        return previousPluginComponents;
      }, {});

      setPluginComponents(updatedComponents);
    };

    if (loadedAllPluginsRef.current) {
      updatePluginsWithNewProps();
    }
  }, [pluggableComponentProps]);

  return pluginComponents;
};
