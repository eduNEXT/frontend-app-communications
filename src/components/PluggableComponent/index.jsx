/* eslint-disable react-hooks/exhaustive-deps */
import React, { useState, useEffect } from 'react';
import loadable from '@loadable/component';
import PropTypes from 'prop-types';

import { isPluginAvailable } from './utils';

/**
 * PluggableComponent - A component that allows dynamic loading and replacement of child components.
 *
 * @param {object} props - Component props
 * @param {React.ReactNode} props.children - Child elements to be passed to the plugin component
 * @param {string} props.as - String indicating the module to import dynamically
 * @param {string} props.id - Identifier for the plugin
 * @param {object} props.pluggableComponentProps - Additional props to be passed to the component
 * @returns {React.ReactNode} - Rendered component
 */
const PluggableComponent = ({
  children,
  as,
  id,
  ...pluggableComponentProps
}) => {
  const [newComponent, setNewComponent] = useState(children);

  useEffect(() => {
    const loadPluginComponent = async () => {
      try {
        const hasModuleInstalled = await isPluginAvailable(as);

        if (hasModuleInstalled) {
          const PluginComponent = loadable(() => import(`@node_modules/@openedx-plugins/${as}`));

          const component = children ? (
            <PluginComponent key={id} {...pluggableComponentProps}>
              {children}
            </PluginComponent>
          ) : (
            <PluginComponent key={id} {...pluggableComponentProps} />
          );

          setNewComponent(component);
        }
      } catch (error) {
        console.error(`Failed to load plugin ${as}:`, error);
      }
    };

    loadPluginComponent();
  }, [id, as, children]);

  return newComponent;
};

PluggableComponent.propTypes = {
  children: PropTypes.node,
  as: PropTypes.string,
  id: PropTypes.string,
};

export default PluggableComponent;
