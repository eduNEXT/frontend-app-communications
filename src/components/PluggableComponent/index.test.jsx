import React, { useState } from 'react';
import {
  render, waitFor, screen, fireEvent,
} from '@testing-library/react';
import PluggableComponent from '.';

const ToggleContentComponent = () => {
  const [showContent, setShowContent] = useState(false);

  return (
    <div>
      <button type="button" onClick={() => setShowContent((prev) => !prev)}>
        Toggle Content
      </button>
      {showContent && <div data-testid="toggle-content">Toggle On</div>}
    </div>
  );
};

describe('PluggableComponent', () => {
  beforeEach(() => {
    jest.resetModules();
  });
  it('renders correctly', async () => {
    const handleClickMock = jest.fn();
    const props = {
      title: 'button title',
      handleClick: handleClickMock,
    };

    const { container } = render(
      <PluggableComponent
        id="pluggableComponent"
        as="communications-app-test-component"
        {...props}
      >
        <h1>Hi this is the original component</h1>
      </PluggableComponent>,
    );

    await waitFor(() => {
      const buttonComponent = screen.getByTestId('button-test');
      expect(buttonComponent).toBeInTheDocument();
      expect(screen.getByText(props.title)).toBeInTheDocument();
      fireEvent.click(buttonComponent);
      expect(handleClickMock).toHaveBeenCalled();
      expect(container).toMatchSnapshot();
    });
  });

  it('loads children component when import is invalid', async () => {
    render(
      <PluggableComponent id="est-pluggable" as="invalid import">
        <div data-testid="plugin">Plugin Loaded</div>
      </PluggableComponent>,
    );

    await waitFor(() => {
      const defaultComponent = screen.getByTestId('plugin');
      expect(screen.getByText('Plugin Loaded')).toBeInTheDocument();
      expect(defaultComponent).toBeInTheDocument();
    });
  });

  it('loads children component when import is empty', async () => {
    render(
      <PluggableComponent
        id="test-pluggable"
        as=""
      >
        <div data-testid="plugin">Plugin Loaded</div>
      </PluggableComponent>,
    );

    await waitFor(() => {
      const defaultComponent = screen.getByTestId('plugin');
      expect(screen.getByText('Plugin Loaded')).toBeInTheDocument();
      expect(defaultComponent).toBeInTheDocument();
    });
  });

  it('returns null when do not have children and import is invalid', async () => {
    render(
      <PluggableComponent
        id="test-pluggable"
        as="invalid-module"
      />,
    );

    await waitFor(() => {
      expect(screen.queryByTestId('plugin')).not.toBeInTheDocument();
    });
  });

  test('updates component when pluggableComponentProps change', async () => {
    const { rerender } = render(
      <PluggableComponent
        id="test-pluggable"
        as="communications-app-test-component"
        title="Testing title component"
      />,
    );

    await waitFor(() => {
      expect(screen.getByText('Testing title component')).toBeInTheDocument();
    });

    rerender(
      <PluggableComponent
        id="test-pluggable"
        as="communications-app-test-component"
        title="Testing a new title component"
      />,
    );

    await waitFor(() => {
      expect(screen.getByText('Testing a new title component')).toBeInTheDocument();
    });
  });

  it('updates component when children change', async () => {
    const { getByText, getByTestId } = render(
      <PluggableComponent
        id="test-pluggable"
        as="default-children"
      >
        <ToggleContentComponent />
      </PluggableComponent>,
    );

    await waitFor(() => {
      const toggleContent = screen.queryByTestId('toggle-content');
      expect(toggleContent).not.toBeInTheDocument();
    });

    const toggleButton = getByText('Toggle Content');
    fireEvent.click(toggleButton);

    // Now, after toggling the content, we expect it to be visible inside PluggableComponent
    await waitFor(() => {
      const toggleContent = getByTestId('toggle-content');
      expect(toggleContent).toBeInTheDocument();
      expect(toggleContent).toHaveTextContent('Toggle On');
    });
  });
});
