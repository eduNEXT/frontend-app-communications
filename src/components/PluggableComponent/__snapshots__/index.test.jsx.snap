// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluggableComponent renders correctly 1`] = `
<div>
  <div
    class="pgn__form-group p-3 border border-success-300"
    data-testid="plugin-input"
  >
    <label
      class="pgn__form-label h3 text-primary-500"
      for="randomID"
    >
      Hello
    </label>
    <div
      class="row container-fluid"
    >
      <div
        class="pgn__form-control-decorator-group col-3"
      >
        <input
          class="form-control is-valid"
          id="randomID"
        />
      </div>
      <p
        class="col-8"
      >
        @openedx-plugins/communications-app-input-form
      </p>
    </div>
    <div
      class="pgn__form-control-description pgn__form-text pgn__form-text-valid"
    >
      <span
        class="pgn__icon"
      >
        <svg
          aria-hidden="true"
          fill="none"
          focusable="false"
          height="24"
          role="img"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2Z"
            fill="currentColor"
          />
        </svg>
      </span>
      <div>
        You are correct
      </div>
    </div>
  </div>
</div>
`;
